### PR TITLE
Session length2

### PR DESCRIPTION
**Pull request by dmose**

### CODE REVIEW

**Comment by msujaws** [comment]
r=me on the changes that we didn't pair together

File: browser/components/loop/content/js/conversation.jsx [hunk 1]
```diff
@@ -120,8 +120,11 @@ loop.conversation = (function(mozL10n) {
     var sdkDriver = new loop.OTSdkDriver({
       isDesktop: true,
       dispatcher: dispatcher,
-      sdk: OT
+      sdk: OT,
+      mozLoop: navigator.mozLoop
     });
+    loop.conversation._sdkDriver = sdkDriver; // expose for functional tests
```
Comment (Standard8) [inline browser/components/loop/content/js/conversation.jsx]: nit: Prefer the comment on the previous line as a full sentence

File: browser/components/loop/content/shared/js/otSdkDriver.js [hunk 2]
```diff
@@ -544,6 +560,72 @@ loop.OTSdkDriver = (function() {
...
+     * @private
+     */
+    _connectionLengthNotedCalls: 0,
+
+    /**
+     * Wrapper for adding a keyed value that also updates
+     * connectionLengthNoted calls and sets this.connectionStartTime to
+     * this.CONNECTION_START_TIME_ALREADY_NOTED.
+     *
+     * @param {number} callLengthSeconds - the call length in seconds
```
Comment (Standard8) [inline browser/components/loop/content/shared/js/otSdkDriver.js]: nit: we don't normally use "-" to separate param names and descriptions. Probably not major, but nice to be consistent.

File: browser/components/loop/test/functional/test_1_browser_call.py [hunk 3]
```diff
@@ -178,13 +243,23 @@ def test_1_browser_call(self):
...
+        self.local_check_media_start_time_initialized()
+
         # XXX To enable this, we either need to navigate the permissions prompt
         # or have a route where we don't need the permissions prompt.
         # self.local_enable_screenshare()
         # self.standalone_check_remote_screenshare()
 
-        # hangup the call
-        self.local_leave_room_and_verify_feedback()
+        # We hangup on the remote side, because this also leaves the
```
Comment (Standard8) [inline browser/components/loop/test/functional/test_1_browser_call.py]: We need a follow-up filed to verify that hanging up on the local side works as well.

File: browser/components/loop/test/functional/test_1_browser_call.py [hunk 4]
```diff
@@ -161,6 +163,66 @@ def local_leave_room_and_verify_feedback(self):
...
+
+            return chatGlobal.''' + expr
+
+        return self.marionette.execute_script(script, [chatbox])
+
+    def local_get_media_start_time(self):
+        return self.local_get_chatbox_window_expr(
+            "loop.conversation._sdkDriver.connectionStartTime")
+
+    # XXX could be memoized
```
Comment (Standard8) [inline browser/components/loop/test/functional/test_1_browser_call.py]: nit: memorized

File: browser/components/loop/test/shared/otSdkDriver_test.js [hunk 5]
```diff
@@ -309,6 +326,93 @@ describe("loop.OTSdkDriver", function () {
 
       sinon.assert.calledOnce(publisher.destroy);
     });
+
+    it("should call _noteConnectionLengthIfNeeded with connection duration", function() {
```
Comment (Standard8) [inline browser/components/loop/test/shared/otSdkDriver_test.js]: We should be checking this in connectionDestroyed and sessionDisconnected sections as well.

**Comment by Standard8** [comment]
I think in general this looks reasonable. However, the functional test feels at the wrong level to me - it feels like its trying to unit/integration test otSdkDriver, rather than functional test the complete process - for a better functional test I'd expect to be looking at the logged telemetry information inside gecko rather than bits of Loop itself.

I think we can handle that in a follow-up though. Please can you file follow-ups for improving the test, and for splitting out the test into separate files etc.

File: browser/components/loop/content/js/conversation.jsx (hunk 1 above)
Comment (dmose) [inline browser/components/loop/content/js/conversation.jsx]: Fixed.

File: browser/components/loop/content/shared/js/otSdkDriver.js (hunk 2 above)
Comment (dmose) [inline browser/components/loop/content/shared/js/otSdkDriver.js]: OK, fixed.  I think I picked up that style from usejsdoc.org, but you're right that we more commonly just use extra whitespace in our code.  Fixed here and elsewhere.

File: browser/components/loop/test/functional/test_1_browser_call.py (hunk 4 above)
Comment (dmose) [inline browser/components/loop/test/functional/test_1_browser_call.py]: Memoized is actually the right word here, though they mean effectively the same thing.  http://en.wikipedia.org/wiki/Memoization

File: browser/components/loop/test/shared/otSdkDriver_test.js (hunk 5 above)
Comment (dmose) [inline browser/components/loop/test/shared/otSdkDriver_test.js]: Great catches; I'd been staring at this all too long.  Fixed!

**Pull request closed without merging**